File: backend/src/main/java/at/ac/tuwien/sepr/groupphase/backend/service/EventService.java
package at.ac.tuwien.sepr.groupphase.backend.service;

import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.event.EventCategoryDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.event.EventDetailDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.event.EventTopTenDto;
import at.ac.tuwien.sepr.groupphase.backend.endpoint.dto.show.ShowDetailDto;
import at.ac.tuwien.sepr.groupphase.backend.entity.Event;
import at.ac.tuwien.sepr.groupphase.backend.entity.Show;
import at.ac.tuwien.sepr.groupphase.backend.exception.ValidationException;
import org.springframework.data.domain.Page;
import org.springframework.data.domain.Pageable;

import java.util.List;

public interface EventService {
    /**
     * Returns the event with the given ID.
     *
     * @param id the ID of the event to retrieve
     * @return the event with the given ID, or null if no such event exists
     */
    Event getEventById(Long id);

    /**
     * Returns all events.
     *
     * @return a list of all events
     */
    List<Event> getAllEvents();

    /**
     * Saves the given event.
     *
     * @param event the event to save
     * @return the saved event
     */
    Event createEvent(Event event) throws ValidationException;

    /**
     * Updates the given event.
     *
     * @param event the event to update
     * @return the updated event
     */
    Event updateEvent(Long id, Event event) throws ValidationException;

    /**
     * Retrieves a paginated list of events associated with the given artist.
     *
     * @param artistId the ID of the artist whose events should be retrieved
     * @param pageable the pagination and sorting information
     * @return a page of {@link EventDetailDto} instances for the given artist
     */
    Page<EventDetailDto> getEventsByArtist(Long artistId, Pageable pageable);

    /**
     * Retrieves a paginated list of shows for a specific event.
     *
     * @param eventId  the ID of the event whose shows should be fetched
     * @param pageable pagination and sorting information
     * @return a {@link Page} of {@link Show} entities linked to the given event
     */
    Page<ShowDetailDto> getPaginatedShowsForEvent(Long eventId, Pageable pageable);

    /**
     * Retrieves a list of the top 10 events by category based on ticket sales.
     *
     * @param category the category of events to filter by
     * @return a list of the top 10 events in the specified category
     */
    List<EventTopTenDto> getTopTenEventsByCategory(String category) throws ValidationException;

    /**
     * Retrieves all event categories.
     *
     * @return a list of all event categories
     */
    List<EventCategoryDto> getAllEventCategories();
}

File: frontend/src/app/services/event.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable, forkJoin } from 'rxjs';
import { Globals } from '../global/globals';
import { CreateEvent } from '../dtos/create-event';
import {Event, EventSearchDto, EventSearchResultDto, EventWithShows} from '../dtos/event';
import {Page} from "../dtos/page";
import { Show } from '../dtos/show';



@Injectable({
  providedIn: 'root'
})
export class EventService {

  private eventBaseUri: string = this.globals.backendUri + '/events';

  constructor(private httpClient: HttpClient, private globals: Globals) {
  }

  /**
   * Creates a new event
   *
   * @param event to create
   */
  create(event: CreateEvent): Observable<Event> {
    console.log('Create event with name: ' + event.name);
    return this.httpClient.post<Event>(this.eventBaseUri, {
      name: event.name,
      category: event.category,
      description: event.description,
      duration: event.duration,
      locationId: event.locationId
    });
  }

  /**
   * Retrieves all events
   */
  getAll(): Observable<Event[]> {
    return this.httpClient.get<Event[]>(this.eventBaseUri);
  }

  /**
   * Retrieves the event with the given ID
   *
   * @param id ID of the event to retrieve
   */
  getEventById(id: number): Observable<Event> {
    return this.httpClient.get<Event>(`${this.eventBaseUri}/${id}`);
  }

  getEventsByIds(ids: number[]): Observable<Event[]> {
    return forkJoin(ids.map(id => this.getEventById(id)));
  }


  /**
   * Retrieves all events linked to a specific artist via shows, paginated.
   *
   * @param artistId ID of the artist whose events should be fetched
   * @param page Page number for pagination (default: 0)
   * @param size Number of items per page (default: 5)
   * @returns An Observable of a paginated list of Event objects
   */
  getEventsByArtist(artistId: number, page = 0, size = 5): Observable<Page<Event>> {
    const url = `${this.eventBaseUri}/by-artist/${artistId}?page=${page}&size=${size}`;
    return this.httpClient.get<Page<Event>>(url);
  }

  /**
   * Retrieves all events specified by the search criteria, paginated.
   *
   * @param criteria Search criteria for filtering events
   * @returns An Observable of a paginated list of EventSearchResultDto objects
   */
  searchEvents(criteria: EventSearchDto): Observable<Page<EventSearchResultDto>> {
    return this.httpClient.post<Page<EventSearchResultDto>>(
      this.eventBaseUri + '/search',
      criteria
    );
  }

  /**
   * Retrieves paginated shows (dates) for a specific event
   *
   * @param eventId ID of the event
   * @param page Page number (default: 0)
   * @param size Page size (default: 5)
   * @returns An Observable of paginated Show objects
   */
  getPaginatedShowsForEvent(eventId: number, page = 0, size = 5): Observable<Page<Show>> {
    const url = `${this.eventBaseUri}/${eventId}/shows/paginated?page=${page}&size=${size}`;
    return this.httpClient.get<Page<Show>>(url);
  }
}
